core/block_svg.js

```diff
@@ -619,9 +619,22 @@ Blockly.BlockSvg.prototype.tab = function(start, forward) {
   var target = list[forward ? i + 1 : i - 1];
   if (!target) {
     // Ran off of list.
-    var parent = this.getParent();
-    if (parent) {
-      parent.tab(this, forward);
+    // If there is an output, tab up to that block.
+    var outputBlock = this.outputConnection && this.outputConnection.targetBlock();
+    if (outputBlock) {
+      outputBlock.tab(this, forward);
+    } else { // Otherwise, go to next / previous block, depending on value of `forward`
+      if (forward) {
+        var nextBlock = this.nextConnection && this.nextConnection.targetBlock();
+        if (nextBlock) {
+          nextBlock.tab(this, forward);
+        }
+      } else {
+        var previousBlock = this.previousConnection && this.previousConnection.targetBlock();
+        if (previousBlock) {
+          previousBlock.tab(this, forward);
+        }
+      }
```

Comment:
At that point, I guess what we have to do is traverse back up the previous connections until we find an input connection, then go to the next connection of that block?